### PR TITLE
fix: unify dense tensor padding convention (dim-0 == batch_size)

### LINKED ISSUE #361: BaseBatch dense tensor shape convention inconsistency with padded fields (num_candidates)
## Problem

`pad_and_allgather_batch` in `batch_allgather.py` assumes all dense tensor fields in `BaseBatch` follow the convention: **dim-0 = `actual_batch_size`**, so it computes `eps = numel // actual_batch_size` to determine the per-sample element count and pads to `batch_size * eps`.

However, some fields like `HSTUBatch.num_candidates` are **already padded to `batch_size` by the dataloader** (via `pad_tensor()` in `hstu_sequence_dataset.py`). When `actual_batch_size < batch_size` (incomplete batch), `numel // actual_batch_size` gives the wrong `eps`, producing an incorrectly sized tensor that causes `all_gather_into_tensor` to hang (size mismatch across ranks).

### Example

- `num_candidates` shape = `[128]` (padded to `batch_size=128`)
- `actual_batch_size = 11`
- `eps = 128 // 11 = 11` (wrong, should be 1)
- `pad_size = 128 * 11 - 128 = 1280`
- Padded tensor = `[1408]` — other ranks send `[128]` → NCCL AllGather deadlock

### Root Cause

There is no clear contract in `BaseBatch` for dense tensor fields:
1. Should dim-0 always equal `actual_batch_size` (unpadded)?
2. Or can dim-0 equal `batch_size` (pre-padded by dataloader)?

Currently both conventions coexist:
- `num_candidates`: pre-padded to `[batch_size]` by dataloader
- Dense labels in test: `[actual_batch_size, num_features]` (unpadded)

This ambiguity also affects:
- `_strip_dense_padding` in `batch_shuffler.py` — uses `t.reshape(full_bs, -1)[:actual_bs].reshape(-1)` which assumes unpadded dim-0
- `BaseBatch.index_select` — uses `t.reshape(self.actual_batch_size, -1)` which assumes dim-0 = actual_batch_size
- KK algorithm and TorchRec require padded KJT lengths (no empty tensors), but the dense tensor convention is undefined

### Workaround

Add a guard before padding: skip if `numel % batch_size == 0` (tensor is already correctly sized):

```python
if pad_dense and tensor_or_kjt.numel() % batch.batch_size != 0:
    eps = _elems_per_sample(tensor_or_kjt, batch.actual_batch_size)
    ...
```

### Suggested Fix

Standardize dense tensor convention in `BaseBatch`:
- Option A: All dense tensors must have dim-0 = `batch_size` (dataloader pads). `pad_and_allgather_batch` and `_strip_dense_padding` skip tensors already at `batch_size`.
- Option B: All dense tensors must have dim-0 = `actual_batch_size` (unpadded). Dataloader must not pad dense fields. `pad_and_allgather_batch` handles all padding.
- Option C: Add a metadata field to `BaseBatch` indicating which dense fields are pre-padded.

### Affected Code

- `examples/commons/distributed/batch_allgather.py` — `allgather_field`, `_pad_dense`
- `examples/commons/distributed/batch_shuffler.py` — `_strip_dense_padding`
- `examples/commons/sequence_batch/batch.py` — `BaseBatch.index_select`
- `examples/commons/datasets/hstu_sequence_dataset.py` — `pad_tensor` for `num_candidates`

### Reproduction

```bash
# 4 GPU movielen ranking with shuffler — hangs at eval's last incomplete batch
cd examples/hstu
torchrun --nproc_per_node 4 --master_addr localhost --master_port 6000   training/pretrain_gr_ranking.py --gin-config-file ./training/configs/movielen_ranking.gin
```

### PR DESCRIPTION
**Pull request by JacoCheung**
## Summary

Fix NCCL AllGather deadlock caused by inconsistent dense tensor padding in `pad_and_allgather_batch`. All dense tensor fields in `BaseBatch` now have dim-0 == `batch_size` (zero-padded by the dataloader), eliminating the ambiguity that caused `_elems_per_sample` to compute wrong padding sizes for pre-padded tensors.

Closes #361

## Changes

- **`batch_allgather.py`**: Remove `_elems_per_sample` and dense padding logic from `allgather_field` — dense tensors are already at `batch_size`, just AllGather directly
- **`batch_all2all.py`**: Remove dense padding logic from `all2all_field` and world_size==1 fast path
- **`batch.py`**: Update `BaseBatch` docstring (dense dim-0 == `batch_size`), fix `index_select_dense_tensor` to use `batch_size` instead of `actual_batch_size`
- **`batch_shuffler.py`**: Update `_strip_dense_padding` docstring
- **`test_distributed.py`**: Update `generate_batch` to pad dense labels to `[batch_size]`

Net: **-108 lines** of padding/reshaping logic removed.

## Root Cause

`pad_and_allgather_batch` assumed dense tensors had `actual_batch_size` elements and computed `eps = numel // actual_batch_size` to determine padding. But `num_candidates` was already padded to `[batch_size]` by the dataloader, so `128 // 11 = 11` (wrong, should be 1), producing a `[1408]` tensor while other ranks sent `[128]` → NCCL size mismatch → deadlock.

## Design Document

# Batch Padding & Shuffle Design

This document describes how padding, shuffling, and load balancing work together in the distributed training pipeline.

## 1. Batch Structure

`BaseBatch` holds two types of data:

| Field Type | Dimension | Padding |
|---|---|---|
| **KJT** (KeyedJaggedTensor) | `batch_size * num_keys` lengths | Padding samples have `length=0` |
| **Dense Tensor** | `batch_size` in dim-0 (flat: `batch_size * eps` elements) | Padding rows are zeros |

**Key invariant**: Both KJT and dense tensors always have `batch_size` in their batch dimension. `actual_batch_size` only records how many leading samples are real (non-padding).

```
batch_size = 128          # full size including padding
actual_batch_size = 100   # first 100 samples are real

KJT lengths:   [L0, L1, ..., L99, 0, 0, ..., 0]   (128 entries, last 28 are zero)
Dense tensor:  [V0, V1, ..., V99, 0, 0, ..., 0]   (128 entries, last 28 are zero)
```

## 2. Why KJT Padding Is Required

KJT lengths **must** be padded to `batch_size` with zeros because:

1. **AllGather symmetry**: `keyed_jagged_tensor_list_allgather` requires all ranks to contribute the same number of length entries.
2. **KK partitioning**: The Karmarkar-Karp algorithm requires `len(workloads) % world_size == 0`. Workloads are derived from KJT lengths, so the global batch must be `batch_size * world_size`.
3. **FBGEMM ops**: `keyed_jagged_index_select_dim1` requires the `batch_size` argument to match the lengths dimension.
4. **Global indexing**: After AllGather, `index_select` uses global indices (0 to `global_batch_size - 1`). Without padding, indices would be invalid.

Padding samples have `length=0`, producing zero values in the KJT. The HSTU attention kernel naturally skips these (zero-length sequences produce no attention output).

## 3. Why Dense Padding Is Required

Dense tensors **must** have dim-0 == `batch_size` because:

1. **AllGather symmetry**: `gather_along_first_dim` / `all_gather_into_tensor` requires all ranks to send the same tensor shape. If one rank has fewer rows, NCCL hangs.
2. **`index_select` correctness**: `BaseBatch.index_select` reshapes dense tensors as `[batch_size, -1]` before indexing. If dim-0 != `batch_size`, the reshape fails.
3. **Consistency with KJT**: After shuffle, `_strip_dense_padding` uses `reshape(batch_size, -1)[:actual_bs]` to extract real rows. This requires dim-0 == `batch_size`.

The dataloader is responsible for padding dense tensors to `batch_size` (e.g., `pad_tensor()` in `hstu_sequence_dataset.py`).

### 4.1 Two-Phase Async Shuffle

The shuffle has two phases, designed to overlap with forward/backward:

```
Phase 1 (on _memcpy_stream, during forward):
  1. Compute per-sample workloads from KJT lengths
  2. AllGather workloads across all ranks
  3. Submit Karmarkar-Karp algorithm to background thread

Phase 2 (on _memcpy_stream, after forward):
  4. Wait for KK result (partition indices)
  5. AllGather batch data (KJT + dense)
  6. index_select to pick this rank's assigned samples
  7. Strip dense padding
```

### 4.2 Karmarkar-Karp (KK) Algorithm

KK partitions the **global batch** (`batch_size * world_size` samples, including padding) into `world_size` equal-size groups, minimizing the max workload across groups.

- Input: `[W0, W1, ..., W_{global_bs-1}]` where padding samples have `W=0`
- Output: `partitions[rank] = [idx_0, idx_1, ..., idx_{batch_size-1}]`
- Each rank gets exactly `batch_size` indices (some may point to padding samples)

### 4.3 `_sort_partitions_padding_last`

After KK, each rank's indices may interleave real and padding samples:

```
Before sort: [3, 17, 0, 12, 8, 15, 2, 20]
                      ^padding     ^padding

After sort:  [3, 17, 12, 8, 2, 20, 0, 15]
              |--- real ---|  |-padding-|
```

This ensures `[:actual_bs]` slicing extracts exactly the real samples, enabling `_strip_dense_padding` to work with a simple slice.

### 4.4 `_strip_dense_padding`

After shuffle + index_select, the batch has `batch_size` samples (some padding). `_strip_dense_padding` removes trailing padding rows from dense tensors:

```python
t.reshape(batch_size, -1)[:actual_bs].reshape(-1)
```

KJT fields are **not** stripped — they keep `batch_size` lengths (with zeros for padding). This is safe because downstream ops (attention, embedding) handle zero-length sequences.

## 5. Communication Pattern

```
AllGather path:
  Local batch [bs] → AllGather → Global batch [bs * W] → index_select [bs] → strip [actual_bs]

All2All path:
  Local batch [bs] → All2All (send to target ranks) → Local batch [bs] → strip [actual_bs]
```

Both paths produce a batch where:
- KJT has `batch_size` samples (padded)
- Dense tensors have `actual_bs` rows (stripped)
- `actual_batch_size` is set to `actual_bs`

## 6. Loss Computation

After shuffle, the model processes `actual_bs` real samples. The loss is computed only on these samples:

```python
# Pipeline handles this via num_loss_tokens():
global_tokens = batch.num_loss_tokens()  # counts only real samples
torch.distributed.all_reduce(global_tokens)
loss = local_loss_sum * dp_size / global_tokens  # normalize by real token count
```

Padding samples either:
- Produce zero loss (because their features/labels are zeros), or
- Are excluded by `_strip_dense_padding` before entering the model

## Test Plan

- [x] `test_distributed.py` — 52 cases (4 GPU), all passed
- [x] `test_batch_shuffler_factory` — passed (4 GPU)
- [x] `pretrain_gr_ranking` 4-GPU movielen integration test — 1000 iters + eval, AUC=0.80, no hang
- [ ] CI pipeline

## CI

- [Pipeline #48488310](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/pipelines/48488310) — 2026-04-14

### CODE REVIEW

**Comment by JacoCheung** [comment]
:x: **Pipeline #48748914** -- failed

| Job | Status | Log |
|-----|--------|-----|
| pre_check | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/299953580) |
| train_build | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/299953581) |
| inference_build | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/299953582) |
| tritonserver_build | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/299953583) |
| build_whl | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/299953584) |
| dynamicemb_test_fwd_bwd_8gpus | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/299953585) |
| dynamicemb_test_load_dump_8gpus | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/299953586) |
| unit_test_1gpu_a100 | :x: failed | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/299953587) |
| unit_test_1gpu_h100 | :x: failed | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/299953588) |
| unit_test_4gpu | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/299953589) |
| unit_test_tp_4gpu | :x: failed | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/299953590) |
| L20_unit_test_1gpu | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/299953591) |
| inference_unit_test_1gpu | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/299953592) |
| inference_test_1gpu | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/299953593) |

**Result: 11/14 jobs passed**

[View full pipeline](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/pipelines/48748914)

**Review by shijieliu — APPROVED** [review]
